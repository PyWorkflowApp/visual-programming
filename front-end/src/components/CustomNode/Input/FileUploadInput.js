import React, { useRef, useState, useEffect } from 'react';
import { Button } from 'react-bootstrap';
import * as API from '../../../API';

/**
 *  Component representing a file parameter.
 *  Uploads selected file to server upon selection, and passes
 *  the filename from the server response to the form callback.
 */
export default function FileUploadInput(props) {

    const input = useRef(null);
    const [fileName, setFileName] = useState(props.value || "");
    const [status, setStatus] = useState(props.value ? "ready" : "unconfigured");

    const {keyName, onChange} = props;
    // fire callback on mount to update node config state
    useEffect(() => {
            onChange(keyName, fileName);
        },
        [fileName, keyName, onChange]);

    const uploadFile = async file => {
        props.disableFunc(true);
        setStatus("loading");
        const fd = new FormData();
        fd.append("file", file);
        fd.append("nodeId", props.node.options.id);
        API.uploadDataFile(fd)
            .then(resp => {
                setFileName(resp.filename);
                setStatus("ready");
                props.disableFunc(false);
                setStatus("ready");
            }).catch(() => {
                setStatus("failed");
            });
        input.current.value = null;
    };
    const onFileSelect = e => {
        e.preventDefault();
        if (!input.current.files) {
          return;
        }
        
        uploadFile(input.current.files[0]);
    };

    if (status === "loading") return (<div>Uploading file...</div>);
    const btnText = status === "ready" ? "Choose Different File" : "Choose File";
    let content;
    if (status === "ready") {
        const rxp = new RegExp(props.node.options.id + '-');
        content = (
            <div>
                <b style={{color: 'green'}}>File loaded:</b>&nbsp;
                {fileName.replace(rxp, '')}
            </div>
        )
    } else if (status === "failed") {
        content = (<div>Upload failed. Try a new file.</div>);
    }
    return (
        <>
            <input type="file" ref={input} onChange={onFileSelect}
                   style={{display: "none"}} />
            <input type="hidden" name={props.name} value={fileName} />
            <Button size="sm" onClick={() => input.current.click()}>{btnText}</Button>
            {content}
        </>
    )
}
