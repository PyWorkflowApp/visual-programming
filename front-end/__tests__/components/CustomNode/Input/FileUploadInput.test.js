import React from 'react'
import { shallow, mount } from 'enzyme';
import FileUploadInput from '../../../../src/components/CustomNode/Input/FileUploadInput';
import CustomNodeModel from '../../../../src/components/CustomNode/CustomNodeModel';

describe('Validates FileUploadInput', () => {

  beforeEach(() => {
    global.console = {log: jest.fn(() => []), error: jest.fn(() => [])};

    global.fetch = jest.fn(() => Promise.resolve({
      ok: true,
      data: [],
      json: jest.fn(() => []),
      text: jest.fn(() => Promise.resolve({})),
      headers:{
        get: (s)=>{
          if (s === "content-type") {
            return "text";
          }

          if (s === "Content-Disposition") {
            return "filenameToDownload";
          }
        }
      }
    }));
  });


  it('Display FileUploadInput', () => {
    const node = new CustomNodeModel({id: "myId", num_in: 2, num_out: 1, is_global: false});
    const onChange = jest.fn(() => []);
    const disableFunc = jest.fn(() => []);
    const input = mount(<FileUploadInput
                      value="ready"
                      disableFunc={disableFunc}
                      onChange={onChange}
                      node={node}
                       />);
    expect(input).toMatchSnapshot();
  });

  it('FileUploadInput selects a file', () => {
    const node = new CustomNodeModel({id: "myId", num_in: 2, num_out: 1, is_global: false});
    const onChange = jest.fn(() => []);
    const disableFunc = jest.fn(() => []);
    const input = mount(<FileUploadInput
                        keyName="uploadFile"
                        disableFunc={disableFunc}
                        onChange={onChange}
                        node={node}
                        />);

    const event = { preventDefault: jest.fn(() => [])};
    input.find({keyName: "uploadFile"}).simulate('change', event);
    expect(event.preventDefault.mock.calls.length).toBe(1);

  });

});
