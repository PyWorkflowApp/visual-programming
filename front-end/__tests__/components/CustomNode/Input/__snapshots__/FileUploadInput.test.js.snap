// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validates FileUploadInput Display FileUploadInput 1`] = `ReactWrapper {}`;
